### PR TITLE
vg/vgimg: change background color to transparent

### PR DESCRIPTION
**Pull request by MiguelPires**
This PR changes vgmimg.go to draw graphs in a transparent background rather than white.

### CODE REVIEW

**Comment by kortschak** [comment]
Is this related to # 473?

You will need to regenerate the golden images.

**Comment by MiguelPires** [comment]
Hi. Yes, it is. Sure, should I just replace all the golden images by the ones generated by the tests?

**Comment by MiguelPires** [comment]
I regenerated the golden images.

**Comment by kortschak** [comment]
You'll need to do that throughout plot. Please run `go test gonum.org/v1/plot/...`.

File: vg/vgimg/vgimg.go [hunk 1]
```diff
@@ -101,7 +101,8 @@ func NewWith(o ...option) *Canvas {
 		c.gc.Scale(1, -1)
 		c.gc.Translate(0, -h)
 	}
-	draw.Draw(c.img, c.img.Bounds(), image.White, image.ZP, draw.Src)
+
+	draw.Draw(c.img, c.img.Bounds(), image.Transparent, image.ZP, draw.Src)
```
Comment (kortschak) [inline vg/vgimg/vgimg.go]: Why did this move?

File: plotter/sankey_test.go [hunk 2]
```diff
@@ -424,14 +424,14 @@ func TestSankey_grouped(t *testing.T) {
 // the input Flows.
 func TestSankey_idempotent(t *testing.T) {
 	flows := []Flow{
-		Flow{
+		{
```
Comment (kortschak) [inline plotter/sankey_test.go]: These unrelated changes should be here.

**Comment by sbinet** [comment]
I don't think we need to regenerate the PDF ones, do we?
only JPG, PNG and TIFF: _i.e._ all the plots generated with the `vgimg` backend.

otherwise, there's some coupling I don't understand :)

**Comment by kortschak** [comment]
@sbinet There are vgimg-generated files throughout, so it's hard to just do those, though a `git add '*.{jpg,tiff,png}'` would have left the others unstaged and those could have then been dropped.

**Comment by MiguelPires** [comment]
I see, I can revert the PDF ones then =) do you want me to?

**Comment by MiguelPires** [comment]
Hi. When you say that maybe there should be a background color option func, do you mean in vgimg.Canvas? I'm not very familiar with this but it seems like the Save function triggers the creation of Canvas objects where the color is set differently. Should the background color be copied over whenever a Canvas object is created?

**Comment by MiguelPires** [comment]
Hi, thank you for the feedback. Do you mean having a background color function at vgimg.Canvas? It seems like the plot.Save function triggers the creation of Canvas objects where the color is unrelated to the plot's background color and draws them. Since the draw happens right after creating the canvas object, how could I use a background color function to solve this?

**Comment by MiguelPires** [comment]
Hi, thank you for the feedback. Do you mean having a background color function at vgimg.Canvas? It seems like the plot.Save function triggers the creation of Canvas objects where the color is unrelated to the plot's background color and draws them. Since the draw happens right after creating the canvas object, how could I use a background color function to solve this?

**Comment by MiguelPires** [comment]
Hi, thank you for the feedback. Do you mean having a background color function at vgimg.Canvas? It seems like the plot.Save function triggers the creation of Canvas objects where the color is unrelated to the plot's background color and draws them. Since the draw happens right after creating the canvas object, how could I use a background color function to solve this?

**Comment by MiguelPires** [comment]
Hi, thank you for the feedback. Do you mean having a background color function at vgimg.Canvas? It seems like the plot.Save function triggers the creation of Canvas objects where the color is unrelated to the plot's background color and draws them. Since the draw happens right after creating the canvas object, how could I use a background color function to solve this?

**Comment by MiguelPires** [comment]
Hi, thank you for the feedback. Do you mean having a background color function at vgimg.Canvas? It seems like the plot.Save function triggers the creation of Canvas objects where the color is unrelated to the plot's background color and draws them. Since the draw happens right after creating the canvas object, how could I use a background color function to solve this?

File: plotter/sankey_test.go (hunk 2 above)
Comment (MiguelPires) [reply plotter/sankey_test.go]: I've reverted them in the latest commit.

File: plotter/testdata/polygon_holes_golden.eps [hunk 3]
```diff
@@ -2,7 +2,7 @@
 %%Creator gonum.org/v1/plot/vg/vgeps
```
Comment (kortschak) [inline plotter/testdata/polygon_holes_golden.eps]: This file can be reverted too.

**Comment by kortschak** [comment]
The duplicates were because GitHub was very very sick yesterday.

**Comment by MiguelPires** [comment]
Thanks for the info kortschak. I've made the changes according to your suggestion and reverted the golden files you mentioned.

File: vg/vgimg/vgimg.go [hunk 4]
```diff
@@ -53,8 +54,8 @@ const (
 )
 
 // New returns a new image canvas.
-func New(w, h vg.Length) *Canvas {
-	return NewWith(UseWH(w, h))
+func New(w, h vg.Length, c color.Color) *Canvas {
```
Comment (kortschak) [inline vg/vgimg/vgimg.go]: I'm not sure this is necessary. @sbinet?

File: vg/vgimg/vgimg.go [hunk 5]
```diff
@@ -29,10 +29,11 @@ import (
...
 type Canvas struct {
-	gc    draw2d.GraphicContext
-	img   draw.Image
-	w, h  vg.Length
-	color []color.Color
+	gc              draw2d.GraphicContext
+	img             draw.Image
+	w, h            vg.Length
+	color           []color.Color
+	backgroundColor color.Color
```
Comment (kortschak) [inline vg/vgimg/vgimg.go]: s/backgroundColor/background/g; the type already tells us it's a colour.

File: vg/vgimg/vgimg.go [hunk 6]
```diff
@@ -165,6 +167,14 @@ func UseImageWithContext(img draw.Image, gc draw2d.GraphicContext) option {
 	}
 }
 
+// UseBackgroundColor specifies a color for the canvas
```
Comment (kortschak) [inline vg/vgimg/vgimg.go]: `// UseBackgroundColor specifies the canvas background color.`

File: vg/draw/canvas.go [hunk 7]
```diff
@@ -269,26 +269,26 @@ func New(c vg.CanvasSizer) Canvas {
 // Supported formats are:
 //
 //  eps, jpg|jpeg, pdf, png, svg, and tif|tiff.
-func NewFormattedCanvas(w, h vg.Length, format string) (vg.CanvasWriterTo, error) {
+func NewFormattedCanvas(w, h vg.Length, color color.Color, format string) (vg.CanvasWriterTo, error) {
```
Comment (kortschak) [inline vg/draw/canvas.go]: s/color/background/g

I'm not wildly happy about additional params here, but I don't see a good alternative.

/cc @sbinet

**Comment by MiguelPires** [comment]
I think the unformatted code is still this https://github.com/gonum/plot/pull/476#discussion_r226270152 . Should I format sankey_test.go?

**Comment by kortschak** [comment]
I've sent a PR to fix that.

File: vg/vgimg/vgimg.go [hunk 8]
```diff
@@ -165,6 +167,14 @@ func UseImageWithContext(img draw.Image, gc draw2d.GraphicContext) option {
 	}
 }
 
+// UseBackgroundColor specifies the canvas background color
```
Comment (kortschak) [inline vg/vgimg/vgimg.go]: Final full stop please.

File: vg/vgimg/vgimg.go (hunk 4 above)
Comment (ctessum) [reply vg/vgimg/vgimg.go]: I don't think this requires an api change.

I think the New function arguments should stay the same, but the body should be:

```
Return NewWith(UseWH(w,h), UseColor(color.White))
```

Then there can be:

``` go
func UseColor(c color.Color) option {
return func(c *Canvas) uint32 {
		draw.Draw(c.img, c.img.Bounds(), image.White, image.ZP, draw.Src)
		return setsColor
	}
}
```

Then we just need to add a setsColor constant and delete the existing draw.Draw line from NewWith.

I'm on a phone now but can clarify later if necessary.

Comment (ctessum) [reply vg/vgimg/vgimg.go]: ...and `UseImg` will need to return `setsSize&setsColor`

Comment (sbinet) [reply vg/vgimg/vgimg.go]: I agree with @ctessum.

File: vg/draw/canvas.go (hunk 7 above)
Comment (sbinet) [reply vg/draw/canvas.go]: for this one... wider changes might be required to regularize canvas modification options across `vg` backends...

perhaps with a new `type` in the `vg` package that `vg` backend can recognize?
```go
package vg

type Option struct {
    Background color.Color
    DPI int
    // others...
}
```

well... this may warrant some more thinking...

Comment (ctessum) [reply vg/draw/canvas.go]: I think there's a tradeoff between the convenience of offering multiple customize-able ways to create canvases and the overhead of maintaining them. I'm not sure what the optimal place in the tradeoff space is, but it might make sense for this convenience function to stay as it is, and people needing customization can use another method to create the canvas.

**Comment by kaythxbye** [comment]
Hi, I am not sure why my small commit suddenly breaks the test. Is there any way to investigate this for me?

**Comment by kortschak** [comment]
That was a network error between travis and other people. I have kicked it.

File: vg/vgimg/vgimg.go (hunk 4 above)
Comment (kortschak) [reply vg/vgimg/vgimg.go]: @MiguelPires Please address this.

File: plotter/plotter.go [hunk 9]
```diff
@@ -12,7 +12,7 @@
 // New* functions return an error if the data contains Inf, NaN, or is
 // empty. Some of the New* functions return other plotter-specific errors
 // too.
-package plotter // import "gonum.org/v1/plot/plotter"
+package plotter
```
Comment (sbinet) [inline plotter/plotter.go]: please leave this import comment.
it makes sure people import "plotter" via the vanity import path `"gonum.org/v1/plot/plotter"`.

File: vg/vg.go [hunk 10]
```diff
@@ -5,7 +5,7 @@
 // Package vg defines an interface for drawing 2D vector graphics.
 // This package is designed with the hope that many different
 // vector graphics back-ends can conform to the interface.
-package vg // import "gonum.org/v1/plot/vg"
+package vg
```
Comment (sbinet) [inline vg/vg.go]: please leave this import comment.
it makes sure people import "vg" via the vanity import path `"gonum.org/v1/plot/vg"`.

File: doc.go [hunk 11]
```diff
@@ -13,4 +13,4 @@
 // gonum.org/v1/plot/plotter package
 // which is documented here:
 // https://godoc.org/gonum.org/v1/plot/plotter
-package plot // import "gonum.org/v1/plot"
+package plot
```
Comment (sbinet) [inline doc.go]: please leave this import comment.
it makes sure people import "plot" via the vanity import path `"gonum.org/v1/plot"`.

**Comment by kaythxbye** [comment]
Oh sorry, I'll revert it.

**Comment by sbinet** [comment]
ping?

File: .github/ISSUE_TEMPLATE [hunk 12]
```diff
@@ -25,7 +25,7 @@ https://groups.google.com/forum/#!forum/gonum-dev
 ### What version of Go and Gonum/plot are you using?
 <!--
 Paste the output of `go version` and if you are installing Gonum/plot from source, paste
-the output of `(cd $(go env GOPATH)/src/gonum.org/v1/plot && git rev-parse HEAD)`.
+the output of `(cd $(go env GOPATH)/src/github.com/blockstatecom/plot && git rev-parse HEAD)`.
```
Comment (kortschak) [inline .github/ISSUE_TEMPLATE]: Please revert all `s!gonum.org/v1!github.com/blockstatecom!g` changes

Comment (kaythxbye) [reply .github/ISSUE_TEMPLATE]: Should be fixed now, sorry again for the mess.

File: vg/draw/canvas.go (hunk 7 above)
Comment (kortschak) [reply vg/draw/canvas.go]: @ctessum So what is your position here? Stay as it is prior to this PR or as it is now in the PR?

File: vg/vgimg/vgimg.go (hunk 4 above)
Comment (kortschak) [reply vg/vgimg/vgimg.go]: Please remove this addition.

File: vg/vgimg/vgimg_test.go [hunk 13]
```diff
@@ -28,7 +29,7 @@ func TestIssue179(t *testing.T) {
 	p.Add(scatter)
 	p.HideAxes()
 
-	c := vgimg.JpegCanvas{Canvas: vgimg.New(5.08*vg.Centimeter, 5.08*vg.Centimeter)}
+	c := vgimg.JpegCanvas{Canvas: vgimg.New(5.08*vg.Centimeter, 5.08*vg.Centimeter, color.White)}
```
Comment (kortschak) [inline vg/vgimg/vgimg_test.go]: And these.

File: vg/draw/canvas.go (hunk 7 above)
Comment (ctessum) [reply vg/draw/canvas.go]: I've filed an alternate PR in #486.

**Comment by ctessum** [comment]
@MiguelPires thanks for getting the ball rolling on this! I just merged #486, which follows your design but with fewer unrelated changes. I'm going to close this, but feel free to open it back up if #486 falls short of what you need.
